src/ucp/rndv/proto_rndv.c

```diff
@@ -762,6 +771,243 @@ void ucp_proto_rndv_receive_start(ucp_worker_h worker, ucp_request_t *recv_req,
+                                &req->send.state.dt_iter.mem_info, sg_count);
+
+    status = UCS_PROFILE_CALL(ucp_proto_request_lookup_proto, worker, ep, req,
+                              proto_select, rkey_cfg_index, &sel_param, length);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucp_trace_req(req,
+                  "%s rva 0x%" PRIx64 " length %zd rreq_id 0x%" PRIx64
+                  " with protocol %s",
+                  ucp_operation_names[op_id], req->send.rndv.remote_address,
+                  length, req->send.rndv.remote_req_id,
+                  req->send.proto_config->proto->name);
+    return status;
+}
+
+static void ucp_proto_rndv_receive_iov_start(ucp_worker_h worker,
+                                             ucp_request_t *recv_req,
+                                             const ucp_rndv_rts_hdr_t *rts,
+                                             const void *buffer, size_t length)
+{
+    void *recv_buffer = recv_req->recv.buffer;
+    ucp_request_t *req;
+    ucp_operation_id_t op_id;
+    uint8_t sg_count;
+    ucp_ep_h ep;
+    ucs_status_t status;
+
+    if (rts->size == 0 || rts->size > recv_req->recv.length ||
+        ucp_datatype_class(recv_req->recv.datatype) != UCP_DATATYPE_CONTIG) {
+        ucp_proto_rndv_receive_start_common(worker, &rts->sreq, recv_req, 0,
+                                            rts->size, buffer, 0);
+        return;
+    }
+
+    UCP_WORKER_GET_VALID_EP_BY_ID(&ep, worker, rts->sreq.ep_id, {
+        ucp_proto_rndv_recv_super_complete_status(recv_req, UCS_ERR_CANCELED);
+        return;
+    }, "RTS on non-existing endpoint");
+
+    req = ucp_request_get(worker);
+    if (req == NULL) {
+        ucs_error("failed to allocate rendezvous reply");
+        return;
+    }
+
+    /* Initialize send request */
+    ucp_proto_request_send_init(req, ep, 0);
+    ucp_proto_rndv_multi_rkey_reset(req);
+    req->send.rndv.remote_req_id = rts->sreq.req_id;
+    req->send.rndv.offset        = 0;
+    ucp_request_set_super(req, recv_req);
+
+    status = ucp_proto_rndv_unpack_iov_rkeys(buffer, length, req);
+    if (status != UCS_OK) {
+        goto release_request;
+    }
+
+    recv_req->status = UCS_OK;
+    UCS_PROFILE_CALL_VOID(ucp_datatype_iter_init_from_dt_state, worker->context,
+                          recv_buffer, rts->size, recv_req->recv.datatype,
+                          &recv_req->recv.state, &req->send.state.dt_iter,
+                          &sg_count);
+
+    op_id  = UCP_OP_ID_RNDV_RECV;
+    status = ucp_proto_rndv_iov_select_proto(worker, req, op_id,
+                                             recv_req->recv.op_attr, 0,
+                                             sg_count);
+    if (status != UCS_OK) {
+        ucp_datatype_iter_cleanup(&req->send.state.dt_iter, UCP_DT_MASK_ALL);
+        goto release_request;
+    }
+    if (req->send.proto_config->proto != &ucp_rndv_get_zcopy_proto) {
```

Comment:
Wouldn't it better to skip all other protos if `rma_count != 0` instead of this fallback logic. It seems little-bit tricky from my point of view. @yosefe @gleon99 WDYT?

Reply:
@ivankochin 
Can we do this in stage 3? The current problem is that it cannot get remote `iov`'s count when try to select proto with algorithm V2. Because the proto select parameter does contains all remote address and `rkey`.